src/components/implementation/tests/micro_xcores/test_ipi_interference.c

```diff
@@ -0,0 +1,253 @@
+#include <stdint.h>
+
+#include "micro_xcores.h"
+
+extern void sched_events_clear(int* rcvd, thdid_t* tid, int* blocked, cycles_t* cycles, tcap_time_t* thd_timeout);
+
+/* Test RCV 2: Close Loop at higher priority => Measure Kernel involvement */
+
+extern unsigned int cyc_per_usec;
+
+static volatile asndcap_t asnd[NUM_CPU] = { 0 };
+static volatile arcvcap_t rcv[NUM_CPU] = { 0 };
+static volatile thdcap_t  thd[NUM_CPU] = { 0 };
+static volatile thdid_t   tid[NUM_CPU] = { 0 };
+static volatile int       blkd[NUM_CPU] = { 0 };
+
+static volatile unsigned long long total_rcvd[NUM_CPU] = { 0 };
+static volatile unsigned long long total_sent[NUM_CPU] = { 0 };
+
+#define MAX_THRS 4
+#define MIN_THRS 1
+
+static volatile thdcap_t  spinner_thd[NUM_CPU] = { 0 };
+static volatile cycles_t  global_time[2] = { 0 };
+
+static volatile int       done_test = 0;
+static volatile int       test_start = 0;
+static volatile int       ret_enable = 1;
+static volatile int       pending_rcv = 0;
+
+static struct             perfdata pd[NUM_CPU] CACHE_ALIGNED;
+
+static void
+test_rcv(arcvcap_t r)
+{
+    int pending = 0, rcvd = 0;
+
+    pending = cos_rcv(r, RCV_ALL_PENDING, &rcvd);
+    assert(pending == 0);
+
+    total_rcvd[cos_cpuid()] += rcvd;
+}
+
+static void
+test_asnd(asndcap_t s)
+{
+    int ret = 0;
+
+    ret = cos_asnd(s, 1);
+    assert(ret == 0 || ret == -EBUSY);
+    if (!ret) total_sent[cos_cpuid()]++;
+}
+
+static void
+test_sync_asnd(void)
+{
+    while (!asnd[TEST_SND_CORE]) ;
+    while (!asnd[TEST_RCV_CORE]) ;
+}
+
+static void
+test_rcv_fn(void *d)
+{
+    arcvcap_t r = rcv[cos_cpuid()];
+    asndcap_t s = asnd[cos_cpuid()];
+
+    while (1) {
+        test_rcv(r);
+        test_asnd(s);
+    }
+}
+
+static void
+test_sched_loop(void)
+{
+    int blocked, rcvd, pending, ret;
+    cycles_t cycles;
+    tcap_time_t timeout, thd_timeout;
+    thdid_t thdid;
+
+    /* Clear Scheduler */
+    sched_events_clear(&rcvd, &thdid, &blocked, &cycles, &thd_timeout);
+
+    while (1) {
+
+        if(cos_cpuid() == TEST_RCV_CORE){
+            do {
+                ret = cos_switch(spinner_thd[cos_cpuid()], BOOT_CAPTBL_SELF_INITTCAP_CPU_BASE, TCAP_PRIO_MAX + 2, 0, 0, 0);
+            } while (ret == -EAGAIN);
+        }
+        while ((pending = cos_sched_rcv(BOOT_CAPTBL_SELF_INITRCV_CPU_BASE, RCV_ALL_PENDING, 0,
+                        &rcvd, &thdid, &blocked, &cycles, &thd_timeout)) >= 0) {
+            if (!thdid) goto done;
+            assert(thdid == tid[cos_cpuid()]);
+            blkd[cos_cpuid()] = blocked;
+done:
+            if (!pending) break;
+        }
+
+        if (blkd[cos_cpuid()] && done_test) return;
+        if (cos_cpuid() == TEST_RCV_CORE && done_test) return;
```

Comment:
Are all of these `return`s OK? Should we not be detecting errors in some way?

Reply:
This `return's only triggered on end of the test. If the test failed then we don't one to block forever on one of the cores just proceed to the end test